Feature4

StanislavDargel opened the pull request.

StanislavDargel committed:
#feature/4
1. ShoppingList Project is Reimported into Maven, created pom.xml file.
2. ProductActualPriceValidationRule.java is deleted.
3. ProductNotRemovedException.java is deleted.
4. In ProductValidationService.java singletone pattern changed into DI pattern.
5. ShoppingListApplication.java starts programm with ProductValidationService injection.
6. Product.java refactored into ProductEntity.java.
7. ProductRepositoryImpl.java fixed method save, find, remove & change parameters. (inMemoryDataBase key String changed on key Long).
8. ValidationService.java fixed method save, remove, delete, change parameters. Added new methods (printProductInformation moved from ConsoleUI.java into ValidationService.java), method normalizeData correct output information (Product Name Letters: First Letter in UpperCase others in LowerCase; Product Price Scale to 2 decimals and Rounding Mode Half Even; Product Discount Scale to 1 decimal, Rounding Mode Floor, Calculate Actual Price).
9. Added packages <productdto>, <mapper>. Created new classes ProductDTO into <productdto>, BeanMapper into <mapper>.
10. Created Test for BeanMapper.java, ProductRepositoryImpl.java, ValidationService.java, ValidationService.java.
11. Test Cover 88% Clasees, 69% methods, 51% lines

StanislavDargel committed:
#feature/4
package <target> addet into gitignore

javagururt commented on src/main/java/com/javaguru/shoppinglist/ShoppingListApplication.java:
```diff
@@ -0,0 +1,15 @@
+class ShoppingListApplication {
+
+    public static void main(String[] args) {
+        ConsoleUI consoleUI = new ConsoleUI(new ValidationService(new ProductRepositoryImpl(), new ProductValidationService(), new BeanMapper()));
```
лучше разделить на несколько строчек, иначе непонятно ничего.

javagururt commented on src/main/java/com/javaguru/shoppinglist/console/ConsoleUI.java:
```diff
@@ -0,0 +1,208 @@
+        while (true) {
+            Scanner scanner = new Scanner(System.in);
+            try {
+                System.out.print("\n1. Create product" +
```
можно переделать на работу с интерфейсом.
Интерфейс `Menu/Action` с методом `execute` например. И далее по принципу ValidationService + Rules.
В этом случае меню можно сделать более элегантным. Сейчас же размазано все и очень много форматирования

javagururt commented on src/main/java/com/javaguru/shoppinglist/console/ConsoleUI.java:
```diff
@@ -0,0 +1,208 @@
+        System.out.print("\nSelect product category: ");
+        Integer caseNum = Integer.parseInt(scanner.nextLine());
+        switch (caseNum) {
+            case 0:
```
Можно заменить. Код станет чуть меньше.
```ProductCategory[] categories = ProductCategory.values();
        if (caseNum < 0 || caseNum >= categories.length) {
            throw new IllegalArgumentException("Category with number " + caseNum + " doesn't exist");
        }
        productDTO.setCategory(categories[caseNum]);

javagururt commented on src/main/java/com/javaguru/shoppinglist/console/ConsoleUI.java:
```diff
@@ -0,0 +1,208 @@
+        System.out.print("\nEdit product price (Y/N)?");
+        System.out.print("\nEnter new product price: ");
+        String priceEditAnswer = scanner.nextLine().toUpperCase();
+        switch (priceEditAnswer) {
```
получение ответа `Y/N` от пользователя в нескольких местах можно вынести в отдельный метод, который будет возвращать `true/false` и тогда можно будет заменить на более короткую запись `if (...) { logic }`

javagururt commented on src/main/java/com/javaguru/shoppinglist/console/ConsoleUI.java:
```diff
@@ -0,0 +1,208 @@
+                break;
+
+        }
+        this.service.printProductInfo(this.service.changeParameters(inputID, foundedProductDTO));
```
`this` лишнее

javagururt commented on src/main/java/com/javaguru/shoppinglist/domain/ProductEntity.java:
```diff
@@ -0,0 +1,95 @@
+    private BigDecimal discount;
+    private String description;
+
+    public ProductEntity() {
```
нет смысла переопределять дефолтный конструктор (пустой) на пустой

javagururt commented on src/main/java/com/javaguru/shoppinglist/repository/ProductRepositoryImpl.java:
```diff
@@ -0,0 +1,69 @@
+
+    @Override
+    public ProductEntity save(ProductEntity productEntity) {
+        for (Map.Entry<Long, ProductEntity> pair : inMemoryDatabase.entrySet()) {
```
это лишнее. Имя не является уникальным идентификатором. Так что если имя совпало это не должен быть апдейт.

javagururt commented on src/main/java/com/javaguru/shoppinglist/service/ValidationService.java:
```diff
@@ -0,0 +1,98 @@
+        if (this.repository.findProductById(id) == null) {
+            throw new ProductNotFoundException("Product with ID " + id + " doesn't exist");
+        }
+        return dataNormalizer(this.beanMapper.toProductDTO(this.repository.findProductById(id)));
```
не увлекайся inline. Лучше разбить несколько вызовов на несколько строк

javagururt commented on src/main/java/com/javaguru/shoppinglist/service/ValidationService.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
equals/hashcode/toString для сервисов/репо и т.д лишнее. Эти методы лучше использовать только для сущностей/дто. Скорее всего ты никогда не будешь сравнивать сервисы :)  да и toString тоже не нужен будет

StanislavDargel committed:
#feature/4
1. In ProductEntity.java deleted an empty constructor;
2. In ProductDTO.java deleted an empty constructor;
3. In ProductRepository.java changed methods on Optional<T>, added new method Optional<ProductEntity> findProductByName();
4. In ProductRepositoryImpl.java deleted next methods: equals, hashCode, toString. ProductEntity save() method fixed, others methods reworked (Optional<T>);
5. In ValidationService.java deleted next methods: equals, hashCode, toString. Inline codes was splitted;
6. In <service/validation> added ProductUniqueValidationRule.java;
7. Redesigned ConsoleUI.java logic: created <actions> package, added
ActionMenu.java, ConsoleUIMenu.java, CreateProduct.java, EditParameters.java, FindProductById.java;

StanislavDargel committed:
#feature/4
1. ConsoleUIMenu.java.deleted;
2. ActionMenu.java, CreateProduct.java, EditParameters.java, FindProductById.java, RemoveProduct.java fixed: void action(Integer actionNum) -> void action();
default boolean isAgreement(String answer) -> default boolean isAgree(String answer). In other Action classes fixed print information.
3. ConsoleUI.java was reworked;
4. In ProductCategory.java removed toString() method, private String categoryName -> private final String categoryName;
5. ProductDiscountValidationRule.java was fixed.
6. ValidationService.java in changeParameters(Long id, ProductDTO productDTO) removed productValidationService.validateProduct(productDTO);
7. ConsoleUI.java removed ConsoleUIMenu.java;
